### PR TITLE
re-implement the `signal_distortion_ratio` metric

### LINKED ISSUE #803: re-implement SDR
## 🚀 Feature

seems we start to have hanging tests with SDR, moreover `fast_bss_eval` is only v0.1
moreover checking some license discrepancies as TM uses apache and BSS is MIT

### Motivation

reduce dependencies :rabbit: 

### Alternatives

keep them only as a reference for testing

### PR DESCRIPTION
**Pull request by quancs**
## What does this PR do?

Fixes #803 

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@quancs this is great! :purple_heart:

File: torchmetrics/functional/audio/sdr.py [hunk 1]
```diff
@@ -12,40 +12,95 @@
...
     toeplitz_conjugate_gradient = None
-    compute_stats = None
-    _normalize = None
-    __doctest_skip__ = ["signal_distortion_ratio"]
 
-from torch import Tensor
 
-from torchmetrics.utilities import rank_zero_warn
-from torchmetrics.utilities.checks import _check_same_shape
+def _symmetric_toeplitz(v: Tensor) -> Tensor:
```
Comment (Borda) [inline torchmetrics/functional/audio/sdr.py:47]: can we have longer var names than one letter?

**Review by Borda — APPROVED** [review]

File: torchmetrics/audio/sdr.py [hunk 2]
```diff
@@ -74,22 +73,14 @@ class SignalDistortionRatio(Metric):
...
 
     .. note::
-       1. when pytorch<1.8.0, numpy will be used to calculate this metric, which causes ``sdr`` to be
-            non-differentiable and slower to calculate
-
-       2. using this metrics requires you to have ``fast-bss-eval`` install. Either install as ``pip install
-          torchmetrics[audio]`` or ``pip install fast-bss-eval``
-
-       3. preds and target need to have the same dtype, otherwise target will be converted to preds' dtype
+       1. preds and target need to have the same dtype, otherwise target will be converted to preds' dtype
```
Comment (SkafteNicki) [inline torchmetrics/audio/sdr.py:76]:
```suggestion
       Preds and target need to have the same dtype, otherwise target will be converted to preds' dtype
```

File: torchmetrics/functional/audio/sdr.py [hunk 3]
```diff
@@ -142,47 +181,43 @@ def signal_distortion_ratio(
...
+    if use_cg_iter is not None and _FAST_BSS_EVAL_AVAILABLE and _TORCH_GREATER_EQUAL_1_8:
         # use preconditioned conjugate gradient
-        sol = toeplitz_conjugate_gradient(acf, xcorr, n_iter=use_cg_iter)
+        sol = toeplitz_conjugate_gradient(R_0, b, n_iter=use_cg_iter)
     else:
+        if use_cg_iter is not None:
+            if _FAST_BSS_EVAL_AVAILABLE is False:
+                warnings.warn(
+                    "The `use_cg_iter` parameter of `SDR` requires that `fast-bss-eval` is installed. "
+                    "To dispear this warning, you could install `fast-bss-eval` using `pip install fast-bss-eval` "
```
Comment (SkafteNicki) [inline torchmetrics/functional/audio/sdr.py:204]:
```suggestion
                    "To make this this warning disappear, you could install `fast-bss-eval` using `pip install fast-bss-eval` "
```

File: torchmetrics/functional/audio/sdr.py [hunk 4]
```diff
@@ -142,47 +181,43 @@ def signal_distortion_ratio(
...
+                    "or set `use_cg_iter=None`. For this time, the solver provided by Pytorch is used.",
+                    UserWarning,
+                )
+            elif _TORCH_GREATER_EQUAL_1_8 is False:
+                warnings.warn(
+                    "The `use_cg_iter` parameter of `SDR` requires a Pytorch version >= 1.8. "
+                    "To dispear this warning, you could change to Pytorch v1.8+ or set `use_cg_iter=None`. "
```
Comment (SkafteNicki) [inline torchmetrics/functional/audio/sdr.py:211]:
```suggestion
                    "To make this this warning disappear, you could change to Pytorch v1.8+ or set `use_cg_iter=None`. "
```

File: torchmetrics/functional/audio/sdr.py [hunk 5]
```diff
@@ -104,28 +156,15 @@ def signal_distortion_ratio(
...
 
     .. note::
-       1. when pytorch<1.8.0, numpy will be used to calculate this metric, which causes ``sdr`` to be
-            non-differentiable and slower to calculate
-
-       2. using this metrics requires you to have ``fast-bss-eval`` install. Either install as ``pip install
-          torchmetrics[audio]`` or ``pip install fast-bss-eval``
-
-       3. preds and target need to have the same dtype, otherwise target will be converted to preds' dtype
+       1. preds and target need to have the same dtype, otherwise target will be converted to preds' dtype
```
Comment (SkafteNicki) [inline torchmetrics/functional/audio/sdr.py:159]:
```suggestion
       Preds and target need to have the same dtype, otherwise target will be converted to preds' dtype
```

**Review by SkafteNicki — COMMENTED** [review]
I assume we can safely remove this check right?
https://github.com/PyTorchLightning/metrics/blob/10b3a8529311579b3db1330842f730a643533065/torchmetrics/audio/sdr.py#L109-L113
else LGTM :]

File: torchmetrics/functional/audio/sdr.py [hunk 6]
```diff
@@ -12,40 +12,95 @@
...
+
+def _compute_autocorr_crosscorr(
+    target: torch.Tensor,
+    preds: torch.Tensor,
+    L: int,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    r"""Compute the auto correlation of `target` and the cross correlation of `target` and `preds` using \ the fast
+    Fourier transform (FFT). \ Let's denotes the symmetric Toeplitz matric of the auto correlation of `target` as
+    `R`, the cross correlation \ as 'b', then solving the equation `Rh=b` could have `h` as the coordinate of
+    `preds` in the column space of \ the L shifts of `target`.
```
[the remark this answers is not in the document]
Comment (quancs) [reply torchmetrics/functional/audio/sdr.py:80]: Yes
```suggestion
    r"""Compute the auto correlation of `target` and the cross correlation of `target` and `preds` using the fast
    Fourier transform (FFT). Let's denotes the symmetric Toeplitz matric of the auto correlation of `target` as
    `R`, the cross correlation as 'b', then solving the equation `Rh=b` could have `h` as the coordinate of
    `preds` in the column space of the L shifts of `target`.
```

File: torchmetrics/functional/audio/sdr.py (hunk 1 above)
Comment (quancs) [reply torchmetrics/functional/audio/sdr.py]: changed

File: torchmetrics/functional/audio/sdr.py [hunk 7]
```diff
@@ -12,40 +12,95 @@
...
+                [1, 0, 1, 2, 3],
+                [2, 1, 0, 1, 2],
+                [3, 2, 1, 0, 1],
+                [4, 3, 2, 1, 0]])
+
+    Returns:
+        a symmetric Toeplitz matrix of shape [..., L, L]
+    """
+    vals = torch.cat([torch.flip(v, dims=(-1,)), v[..., 1:]], dim=-1)
+    L = v.shape[-1]
```
[the remark this answers is not in the document]
Comment (quancs) [reply torchmetrics/functional/audio/sdr.py]: changed

**Comment by quancs** [comment]
> I assume we can safely remove this check right?
> 
> https://github.com/PyTorchLightning/metrics/blob/10b3a8529311579b3db1330842f730a643533065/torchmetrics/audio/sdr.py#L109-L113
> 
> else LGTM :]

Removed

**Review by SkafteNicki — APPROVED** [review]

File: torchmetrics/functional/audio/sdr.py [hunk 8]
```diff
@@ -12,40 +12,97 @@
...
+    v_len = vector.shape[-1]
+    return torch.as_strided(
+        vec_exp, size=vec_exp.shape[:-1] + (v_len, v_len), stride=vec_exp.stride()[:-1] + (1, 1)
+    ).flip(dims=(-1,))
+
+
+def _compute_autocorr_crosscorr(
+    target: torch.Tensor,
+    preds: torch.Tensor,
+    L: int,
```
Comment (Borda) [inline torchmetrics/functional/audio/sdr.py:77]: pls rename also here

File: torchmetrics/functional/audio/sdr.py [hunk 9]
```diff
@@ -12,40 +12,97 @@
...
+    # computes the auto correlation of `target`
+    T = torch.fft.rfft(target, n=n_fft, dim=-1)
+    # R_0 is the first row of the symmetric Toeplitz matric
+    R_0 = torch.fft.irfft(T.real**2 + T.imag**2, n=n_fft)[..., :L]
+
+    # computes the cross-correlation of `target` and `preds`
+    P = torch.fft.rfft(preds, n=n_fft, dim=-1)
+    TP = T.conj() * P
+    b = torch.fft.irfft(TP, n=n_fft, dim=-1)[..., :L]
+    return R_0, b
```
Comment (Borda) [inline torchmetrics/functional/audio/sdr.py:105]: the same capital in all PR, pls

File: torchmetrics/functional/audio/sdr.py [hunk 10]
```diff
@@ -12,40 +12,96 @@
...
-        from fast_bss_eval.numpy.linalg import toeplitz
-        from fast_bss_eval.numpy.metrics import compute_stats
+# import or redirect the solve function
+if _TORCH_GREATER_EQUAL_1_8:
+    solve = torch.linalg.solve
+else:
+    from torch import solve as _solve
 
-        solve = numpy.linalg.solve
+    def solve(A: Tensor, b: Tensor) -> Tensor:
```
Comment (quancs) [inline torchmetrics/functional/audio/sdr.py:37]: @Borda 
I guess `A` is OK here right? as it's `A` in pytorch `solve`. https://pytorch.org/docs/1.5.0/torch.html#torch.solve

File: torchmetrics/functional/audio/sdr.py (hunk 9 above)
Comment (quancs) [reply torchmetrics/functional/audio/sdr.py]: changed

File: torchmetrics/functional/audio/sdr.py (hunk 8 above)
Comment (quancs) [reply torchmetrics/functional/audio/sdr.py]: changed

File: torchmetrics/functional/audio/sdr.py (hunk 10 above)
Comment (Borda) [reply torchmetrics/functional/audio/sdr.py:37]: yep, this shall be fine :)
